### PR TITLE
Improved JSON Schema not to allow fields that are not typical for concrete tool type

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
This PR improves JSON Schema not to allow fields that are not typical for concrete tool type.
Previously it was done by `DevfileIntegrityValidator` but actually, it is not a thing that should be checked by integrity validator and schema is able to handle it.

Also, it contains the following small but important changes:
- Check if messages are equal instead of matching pattern in DevfileSchemaValidatorTest.
It allows avoiding characters escaping for improving test cases readability.

- Format expected messages in DevfileSchemaValidatorTest for improving test cases readability.

### What issues does this PR fix or reference?
This PR DOES NOT RESOLVE but is related to https://github.com/eclipse/che/issues/12587

#### Release Notes
N/A

#### Docs PR
N/A

### CODE REVIEW

**Review by skabashnyuk — DISMISSED** [review]

**Comment by mshaposhnik** [comment]
I am generally -1 on this. Looks like this is an unspecified trick, and/or you're operate on bug or side effect. This can potentially cause indefinite behavior. In fact you re-define tool properties with an empty object, when condition matches.  

I wold rather propose to make more researches on that, and maybe we find solution using  "not" operators, etc. Several promising example proposals can be found here:
https://stackoverflow.com/questions/30515253/json-schema-valid-if-object-does-not-contain-a-particular-property/30675218

**Comment by sleshchenko** [comment]
> Looks like this is an unspecified trick, and/or you're operate on bug or side effect. This can potentially cause indefinite behavior. In fact you re-define tool properties with an empty object, when condition matches.

It depends on how you understand Devfile Schema. I do not re-define tool properties. Tools properties are defined in root schema, and tool type specific subschema only guarantees that there is no indistinctive fields for this particular tool type.
So, you do know any issues that can be caused by these changes?

I'm ok with finding alternative ways but I don't like defining like 
```json
          {
            "properties": {
              "type": {
                "enum": [
                  "cheEditor",
                  "chePlugin"
                ]
              }
            },
            "required": [
              "id"
            ],
            "not": {
                "required": [
                     //dockerimage
                     "name",
                     "image",
                     "memoryLimit",
                     "mountSources",
                     "volumes",
                     "env",
                     "endpoints",
                    //kubernetes/openshift
                    "local",
                    "selector"
              ]
            }
          }
```
Because then each tool type subschema should define fields of other tool types specific fields.

**Comment by garagatyi** [comment]
@mshaposhnik @sleshchenko I'm not experienced in JSON Schema enough to decide with whom I agree. Could you provide an explanation (maybe links to schema specification section) that shows whether this trick is legal for the schema or not? Right now you both have your own opinion on that but didn't provide any background on why you have this opinion.

**Comment by garagatyi** [comment]
@sleshchenko your PR description would effectively close mentioned issue, see github hint on that. Please change the working to avoid unintended closing of the issue

**Comment by sleshchenko** [comment]
> @sleshchenko your PR description would effectively close mentioned issue, see github hint on that. Please change the working to avoid unintended closing of the issue

Good catch. Fixed.

**Comment by sleshchenko** [comment]
From https://json-schema.org/
screenshot_20190220_102040

I think validators MUST support empty objects for properties definition and no indefinite behavior should be expected.
@mshaposhnik Could you please provide a result of your investigation and playing with defining the same properties in different subschemas?

**Comment by mshaposhnik** [comment]
There is no problem with empty objects at all. Problem that we duplicate definitions.
From my POW adding duplicate definition for the same property (even if this definition is empty) is a not safe trick, and how validator behave is depends on its internal implementation (it may override definitions or merge them or etc). Schema itself has no examples with definition the same prorerty twice (even in different subschemas).  Let's also not forget that our schema is intended to be public,  and such tricks may be confusing.

**Comment by sleshchenko** [comment]
I expect that subschemas are validated separately. And properties are not duplicated at all. Sorry, but I don't understand what the problem if two subschemas define the same properties.

**Comment by mshaposhnik** [comment]
Even if they validated separately, then by logic, the internal emtpy definition `"local": {}` should allow `local` to be any type of object without resctictions (which is fact not true, it is still validated using it's external definition, e.g. you cannot send integer if it's described as string in main defilition).

**Comment by sleshchenko** [comment]
@mshaposhnik And what about how we define `type` property? Do you think it's tricky too? I mean we define in root schema that it's an string https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json#L161
and then in each of `oneOf` subschemas we duplicate definition for `type` property https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json#L115

**Comment by mshaposhnik** [comment]
This is an validation criteria - it defines nothing about `type` object itself (and it is directly described in specification to be such).

**Review by mshaposhnik — APPROVED** [review]
Discussed again and agreed that it's ok as a _temporary_ solution until we can migrate to 07 schema version and implement it in other way.
Please check that it's not hurts docs generation much before merge.

**Comment by sleshchenko** [comment]
@mshaposhnik Here is how my changes affect Devfile Docs:
Diff - https://github.com/sleshchenko/devfile/commit/48899a047b1eff6183eea84bfec365de97d6bcf9
index.md - https://github.com/sleshchenko/devfile/blob/checkDocsAfterDevfileChanges/docs/index.md
It looks OK for me. Could you please review it?

**Comment by mshaposhnik** [comment]
LGTM

**Comment by sleshchenko** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

File: wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/validator/DevfileSchemaValidatorTest.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ public void setUp() throws Exception {
   @Test(dataProvider = "validDevfiles")
   public void shouldDoNotThrowExceptionOnValidationValidDevfile(String resourceFilePath)
       throws Exception {
-    schemaValidator.validateBySchema(getResource(resourceFilePath), false);
+    schemaValidator.validateBySchema(getResource(resourceFilePath), true);
```
Comment (garagatyi) [inline wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/validator/DevfileSchemaValidatorTest.java]: Are you gonna leave true here?

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json [hunk 2]
```diff
@@ -130,24 +133,36 @@
                   "kubernetes",
                   "openshift"
                 ]
-              }
+              },
+              "name": {},
+              "local": {},
```
Comment (garagatyi) [inline wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json:143]: localContent is also needed to be added

**Review by garagatyi — COMMENTED** [review]
Looks good, but needs syncing with last changes in devfile.json I merged recently

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json (hunk 2 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json:143]: Added

File: wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/validator/DevfileSchemaValidatorTest.java (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/validator/DevfileSchemaValidatorTest.java]: is not actual anymore

**Comment by sleshchenko** [comment]
@mshaposhnik @garagatyi Please take a look one more time after rebasing with the master branch. Note that two more commits are added except JSON Schema improvements. Their descriptions are listed in the PR description.

**Review by mshaposhnik — APPROVED** [review]

**Review by garagatyi — APPROVED** [review]

**Comment by metlos** [comment]
I hate to be late to the discussion, but have you tried defining all the attributes in definitions and referencing them in the individual properties of the various tool types using `$ref`? While using both `oneOf` and `properties` in the tools definition is very clever, it is a little bit hard to get one's head around. At least for me :smile:

**Comment by sleshchenko** [comment]
> have you tried defining all the attributes in definitions and referencing them in the individual properties of the various tool types using `$ref`? While using both `oneOf` and `properties` in the tools definition is very clever, it is a little bit hard to get one's head around. At least for me

I have not tried it. Not sure if it will improve Devfile JSON Schema readability but it would make things more obvious at least until our schema validator/POJO generator does not support if/else statement.

**Pull request merged**